File: client_leger/lib/services/http_service.dart
import 'dart:convert';

import 'package:http/http.dart' as http;
import 'package:http_parser/http_parser.dart';

import '../classes/game.dart';
import '../config/environment.dart';

const String authUrl = 'auth';
const String userInfoUrl = 'userInfo';
const String opponentInfoUrl = 'opponentInfo';
const String avatarUrl = 'images/avatars';
const String badgeUrl = 'images/badges';
const String baseUrl = environment;
const String CLOUD_NAME = 'dejrgre8q';
String url = getServerURL();

class HttpService {
  HttpService();
  Future<http.Response> loginUser(String username) {
    String loginUrl = '$url/api/$authUrl/login';
    return http.put(Uri.parse(loginUrl), body: {"username": username});
  }

  Future<http.Response> logoutUser(String username) {
    String loginUrl = '$url/api/$authUrl/logout';
    return http.put(Uri.parse(loginUrl), body: {"username": username});
  }

  Future<http.Response> loginUserLogs(String email) {
    return http.get(Uri.parse('$url/api/$authUrl/user/$email'));
  }

  Future<http.Response> logoutUserLogs(String username) {
    String logoutUrl = '$url/api/$authUrl/logout';
    return http.put(Uri.parse(logoutUrl), body: {"username": username});
  }

  Future<http.Response> getUserInfo(String email) {
    return http.get(Uri.parse('$url/api/userInfo/$email'));
  }

  Future<http.Response> getStatsInfo(String email) {
    return http.get(Uri.parse('$url/api/stats/$email'));
  }

  //  http://localhost:3000/api/stats/byUsername/Tounsi
  Future<http.Response> getOtherUserProfileInfo(String username) {
    return http.get(Uri.parse('$url/api/stats/byUsername/$username'));
  }

  Future<http.Response> createUser(String email, String username) {
    return http.post(Uri.parse('$url/api/$authUrl/user'),
        body: {"email": email, "username": username});
  }

  Future<http.Response> resetUserPassword(String email) {
    return http.post(Uri.parse('$url/api/$authUrl/user/reset/$email'),
        body: {"email": email});
  }

  Future<http.Response> updateUserSettings(
      String userEmail, Account clientAccountInfo) {
    final Map<String, dynamic> body = {
      'username': clientAccountInfo.username,
      'email': clientAccountInfo.email,
      'userSettings': {
        'avatarUrl': clientAccountInfo.userSettings.avatarUrl,
        'defaultLanguage': clientAccountInfo.userSettings.defaultLanguage,
        'defaultTheme': clientAccountInfo.userSettings.defaultTheme,
        'victoryMusic': clientAccountInfo.userSettings.victoryMusic,
      },
      'progressInfo': {
        'totalXP': clientAccountInfo.progressInfo.totalXP,
        'currentLevel': clientAccountInfo.progressInfo.currentLevel,
        'currentLevelXp': clientAccountInfo.progressInfo.currentLevelXp,
        'xpForNextLevel': clientAccountInfo.progressInfo.xpForNextLevel,
        'victoriesCount': clientAccountInfo.progressInfo.victoriesCount
      },
      'totalXP': clientAccountInfo.progressInfo.totalXP,
      'highScores': clientAccountInfo.highScores,
      'badges': clientAccountInfo.badges,
      'bestGames': clientAccountInfo.bestGames,
      'gamesPlayed': clientAccountInfo.gamesPlayed,
      'gamesWon': clientAccountInfo.gamesWon,
    };

    return http.patch(Uri.parse('${url}/api/${userInfoUrl}/${userEmail}'),
        headers: <String, String>{
          'Content-Type': 'application/json; charset=UTF-8',
        },
        body: jsonEncode(body));
  }

  Future<http.Response> getCloudinarySignature() {
    return http
        .get(Uri.parse('$url/api/images/signature'))
        .catchError((e) => print(e));
  }

  Future<http.Response> uploadFile(Map<dynamic, dynamic> data) async {
    var stream = http.ByteStream(Stream.castFrom(data['file'].openRead()));
    var length = await data['file'].length();

    var uri =
        Uri.parse('https://api.cloudinary.com/v1_1/${CLOUD_NAME}/auto/upload');

    var request = http.MultipartRequest("POST", uri);
    var multipartFile = http.MultipartFile('file', stream, length,
        filename: data['file'].path.split('/').last,
        contentType: MediaType('image', 'jpeg'));
    request.files.add(multipartFile);
    request.fields.addAll(<String, String>{
      'signature': data['signature'],
      'timestamp': data['timestamp'],
      'api_key': data['api_key']
    });

    final res = await request.send();
    return http.Response.fromStream(res);
  }
}

File: client/src/app/services/http.service.ts
import { HttpClient, HttpErrorResponse, HttpHeaders, HttpResponse, HttpStatusCode } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Score } from '@app/classes/score';
import {
    DEFAULT_MESSAGE,
    FORBIDDEN_MESSAGE,
    GONE_RESSOURCE_MESSAGE,
    UNREACHABLE_SERVER_MESSAGE,
    // eslint-disable-next-line prettier/prettier
    UNREACHABLE_SERVER_STATUS_CDOE,
} from '@app/constants/http-constants';
import { PlayerGameStats } from '@app/constants/player-stats';
import { Game } from '@app/interfaces/game';
import { ClientAccountInfo } from '@app/interfaces/serveur info exchange/client-account-info';
import { Observable, of } from 'rxjs';
import { catchError } from 'rxjs/operators';
import { environment } from 'src/environments/environment';

const CLOUD_NAME = 'dejrgre8q';

@Injectable({
    providedIn: 'root',
})
export class HttpService {
    private readonly scoresBaseUrl: string;
    private readonly gameBaseUrl: string;
    private readonly authUrl: string;
    private readonly userInfoUrl: string;
    private readonly opponentInfoUrl: string;
    private readonly avatarUrl: string;
    private readonly badgeUrl: string;
    private readonly statsUrl: string;
    private baseUrl: string;
    private errorMessage: string;
    constructor(private http: HttpClient) {
        this.scoresBaseUrl = 'scores';
        this.gameBaseUrl = 'games';
        this.authUrl = 'auth';
        this.userInfoUrl = 'userInfo';
        this.opponentInfoUrl = 'opponentInfo';
        this.avatarUrl = 'images/avatars';
        this.badgeUrl = 'images/badges';
        this.statsUrl = 'stats';
        this.baseUrl = environment.serverUrl;
        this.errorMessage = '';
    }
    getPlayerStats(playerEmail: string): Observable<PlayerGameStats> {
        return this.http
            .get<PlayerGameStats>(`${this.baseUrl}/${this.statsUrl}/${playerEmail}`)
            .pipe(catchError(this.handleError<PlayerGameStats>('Could not get stats')));
    }
    getCloudinarySignature(): Observable<{ timestamp: string; signature: string; apiKey: string }> {
        return this.http
            .get<{ timestamp: string; signature: string; apiKey: string }>(`${this.baseUrl}/images/signature`)
            .pipe(catchError(this.handleError<{ timestamp: string; signature: string; apiKey: string }>('getPredefinedAvatarsURL')));
    }

    getOpponentInfo(username: string): Observable<ClientAccountInfo> {
        // this.clearError();
        return this.http.get<ClientAccountInfo>(`${this.baseUrl}/${this.userInfoUrl}/${this.opponentInfoUrl}/${username}`);
    }

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    uploadFile(data: FormData): Observable<any> {
        return this.http.post(`https://api.cloudinary.com/v1_1/${CLOUD_NAME}/auto/upload`, data);
    }
    updateUserSettings(userEmail: string, newClientAccountInfo: ClientAccountInfo): Observable<ClientAccountInfo> {
        this.clearError();

        return this.http.patch<ClientAccountInfo>(`${this.baseUrl}/${this.userInfoUrl}/${userEmail}`, newClientAccountInfo);
    }
    getUsernames() {
        this.clearError();
        return this.http.get<string[]>(`${this.baseUrl}/${this.authUrl}/usernames`).pipe(catchError(this.handleError<string[]>('getUsers')));
    }
    getUserInfo(email: string) {
        this.clearError();
        return this.http.get<ClientAccountInfo>(`${this.baseUrl}/${this.userInfoUrl}/${email}`);
    }

    getPredefinedAvatarsURL(): Observable<string[]> {
        this.clearError();
        return this.http.get<string[]>(`${this.baseUrl}/${this.avatarUrl}`).pipe(catchError(this.handleError<string[]>('getPredefinedAvatarsURL')));
    }

    getAvatarURL(imageId: string) {
        this.clearError();
        return `${this.baseUrl}/${this.avatarUrl}/${imageId}`;
    }
    getBadgeURL(badgeId: string) {
        this.clearError();
        return `${this.baseUrl}/${this.badgeUrl}/${badgeId}`;
    }
    loginUser(email: string) {
        this.clearError();
        return this.http.get<unknown[]>(`${this.baseUrl}/${this.authUrl}/user/${email}`).pipe(catchError(this.handleError<string[]>('loginUser')));
    }

    resetUserPassword(email: string) {
        this.clearError();
        return this.http
            .post<{ email: string }>(
                `${this.baseUrl}/${this.authUrl}/user/reset/${email}`,
                { email },
                { headers: this.createCacheHeaders(), observe: 'response' },
            )
            .pipe(catchError(this.handleError<HttpResponse<string[]>>('resetUserPassword')));
    }

    signUpUser(email: string, username?: string) {
        this.clearError();
        // TODO: Connect to server user creation route
        return this.http
            .post<{ email: string; username: string }>(
                `${this.baseUrl}/${this.authUrl}/user`,
                { email, username },
                {
                    headers: this.createCacheHeaders(),
                    observe: 'response',
                },
            )
            .pipe(catchError(this.handleError<HttpResponse<{ email: string; username: string }>>('SignUp User')));
    }

    logoutUser(username: string) {
        this.clearError();
        const logoutUrl = `${this.baseUrl}/${this.authUrl}/logout`;
        return this.http
            .put<string>(logoutUrl, { username }, { headers: this.createCacheHeaders(), observe: 'response' })
            .pipe(catchError(this.handleError<HttpResponse<string>>('logoutUser')));
    }

    fetchAllScores(): Observable<Score[]> {
        return this.http.get<Score[]>(`${this.baseUrl}/${this.scoresBaseUrl}`).pipe(catchError(this.handleError<Score[]>('fetchAllScores')));
    }

    getNScoresByCategory(gameType: string, quantity: number): Observable<Score[]> {
        this.clearError();
        const scoresUrl = `${this.baseUrl}/${this.scoresBaseUrl}/game-type/${gameType}?quantity=${quantity}`;
        return this.http
            .get<Score[]>(scoresUrl, { headers: this.createCacheHeaders() })
            .pipe(catchError(this.handleError<Score[]>('getNScoresByCategory')));
    }
    getAllScoresByCategory(gameType: string): Observable<Score[]> {
        this.clearError();
        const scoresUrl = `${this.baseUrl}/${this.scoresBaseUrl}/game-type/${gameType}`;
        return this.http
            .get<Score[]>(scoresUrl, { headers: this.createCacheHeaders() })
            .pipe(catchError(this.handleError<Score[]>('getAllScoresByCategory')));
    }
    reinitializeScores(): Observable<void> {
        this.clearError();
        const scoresUrl = `${this.baseUrl}/${this.scoresBaseUrl}`;
        return this.http
            .delete<void>(scoresUrl, { headers: this.createCacheHeaders() })
            .pipe(catchError(this.handleError<void>('reinitializeScores')));
    }
    reinitializeHistory(): Observable<void> {
        this.clearError();
        const gamesUrl = `${this.baseUrl}/${this.gameBaseUrl}`;
        return this.http.delete<void>(gamesUrl, { headers: this.createCacheHeaders() }).pipe(catchError(this.handleError<void>('deleteGames')));
    }

    getAllGames() {
        this.clearError();
        const httpHeaders = new HttpHeaders().set('Cache-Control', 'no-cache').set('Expires', '0');
        const gameUrl = `${this.baseUrl}/${this.gameBaseUrl}`;
        return this.http.get<Game[]>(gameUrl, { headers: httpHeaders }).pipe(catchError(this.handleError<Game[]>('getAllGames')));
    }
    anErrorOccurred(): boolean {
        return this.errorMessage !== '';
    }
    getErrorMessage(): string {
        return this.errorMessage;
    }
    private createCacheHeaders(): HttpHeaders {
        return new HttpHeaders().set('Cache-Control', 'no-cache').set('Expires', '0');
    }
    private clearError() {
        this.errorMessage = '';
    }
    private handleErrorStatusCode(code: number) {
        switch (code) {
            case UNREACHABLE_SERVER_STATUS_CDOE:
                this.errorMessage = UNREACHABLE_SERVER_MESSAGE;
                break;
            case HttpStatusCode.Gone:
                this.errorMessage = GONE_RESSOURCE_MESSAGE;
                break;
            case HttpStatusCode.Forbidden:
                this.errorMessage = FORBIDDEN_MESSAGE;
                break;
            default:
                this.errorMessage = DEFAULT_MESSAGE;
                break;
        }
    }
    private handleError<T>(request: string, result?: T): (error: HttpErrorResponse) => Observable<T> {
        return (error: HttpErrorResponse): Observable<T> => {
            this.handleErrorStatusCode(error.status);
            return of(result as T);
        };
    }
}
